firedrake/locate.c

```diff
@@ -25,9 +32,36 @@ int locate_cell(struct Function *f,
         if (f->extruded == 0) {
             for (uint64_t i = 0; i < nids; i++) {
                 if ((*try_candidate)(data_, f, ids[i], x)) {
+                    /* Found cell! */
                     cell = ids[i];
                     break;
                 }
+                else {
+                    /* Cell not found, but could be on cell boundary. We therefore look
+                       at our reference coordinates and find the point closest to being
+                       inside the reference cell. If we don't find a cell using try_candidate
+                       we assume that this process has found our cell. */
+                    /* First we must find the the reference coordinate which is closest to
+                       the interval [0, 1] without being in it (if all reference coordinates
+                       were in the interval we would have already found our cell!). We do
+                       this by storing the distance of each coordinate dimension from the
+                       interval. Note that ref_coords was updated as data_ by try_candidate. */
+                    current_closest_ref_coord = -1.0;
+                    for (uint64_t j = 0; j < dim; j++) {
+                        if(ref_coords->X[j] < 0.0 && current_closest_ref_coord < ref_coords->X[j]) {
+                            current_closest_ref_coord = ref_coords->X[j];
+                        }
+                        else if(1.0 - ref_coords->X[j] < 0.0 && current_closest_ref_coord < 1.0 - ref_coords->X[j]) {
+                            current_closest_ref_coord = 1.0 - ref_coords->X[j];
+                        }
+                    }
+                    /* Then, if it's lower than the last, we assume we must have found our
+                       cell until told otherwise */
+                    if(current_closest_ref_coord < closest_ref_coord){
```

Comment:
`current_closest_ref_coord > closest_ref_coord`?

The point is closer to the current cell if `current_closest_ref_coord` is "less negative".

Reply:
Good spot - clearly been a while since I wrote any code! Fixed